plugins/notifications-backend/src/service/router.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { fullFormats } from 'ajv-formats/dist/formats';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Context, OpenAPIBackend, Request } from 'openapi-backend';
+
+import { Paths } from '../openapi';
+import { checkUserPermission } from './auth';
+import { initDB } from './db';
+import {
+  createNotification,
+  getNotifications,
+  getNotificationsCount,
+  setRead,
+} from './handlers';
+import {
+  notificationsCreatePermission,
+  notificationsReadPermission,
+  notificationsSetReadPermission,
+} from './permissions';
+import { RouterOptions } from './types';
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { logger, dbConfig, catalogClient } = options;
+
+  // create DB client and tables
+  if (!dbConfig) {
+    logger.error('Missing dbConfig');
+    throw new Error('Missing database config');
+  }
+
+  const dbClient = await initDB(dbConfig);
+
+  // create openapi requests handler
+  const api = new OpenAPIBackend({
+    ajvOpts: {
+      formats: fullFormats, // open issue: https://github.com/openapistack/openapi-backend/issues/280
+    },
+    validate: true,
+    definition: '../../plugins/notifications-backend/src/openapi.yaml',
+  });
+
+  await api.init();
+
+  api.register(
```

Comment:
Still not sure it's good idea to allow creation of notifications through REST API. I would prefer just doing it from the code by other plugins (like in the alternative implementation PR)

Reply:
It is required that external systems can emit  new messages to users (primary case). Or maybe even other BE plugins (secondary). Do we have a better option?

Reply:
I think depends on what course we want to take here. If we could block the notifications from direct API calls to the notifications plugin and allow only other BE plugins to create notifications, it would be more safe for the backstage core. You could then create your own simple BE plugin that allows notifications from external callers and there you could do the magic JWT/header permission check things. Or even fetch notifications from some other database or service periodically. This would already be possible with this https://github.com/backstage/backstage/pull/21889

Reply:
we provide API for any source in any language. Besides, even the FEs and BEs are in different processes, are they not?

Reply:
A specific scenario to support the need for external invocation of the API: 
When async-task is initiated from backstage by a user via one of the plugins, e.g. for invoking a workflow/aysnc-action.
During that task, there might be a need for the user to provide additional input or take some action, based on the workflow progress.

Sending a notification by the external system with the required information to the user simplifies the flow of providing additional info or taking the required action by the user and mainly increases user awareness for the step needs to be taken.

Having had to implement by each plugin a mediation between the business logic of the plugin to the notification plugin introduces a complexity in design that could be avoided.

Reply:
I don't say there's no need for external systems to be able to send notifications, I am just pointing out that having custom headers and authentication for this is not the best way to go. That's why separating the notification sending to backend plugins would allow other BE plugins to do custom authentication if needed. Having a different BE doing this would be the application developer's responsibility instead of having such auth magic in the upstream. Processes don't matter as the notifications live in a shared database. 

Also, you don't need to implement this for every plugin if you don't want to. You could just create a very simple rest API that allows the custom authentication and then just uses the notification service to do the actual sending.

Reply:
as far as I know, the auth implementation is not specific to this plugin. it is the backstage auth

Reply:
But it is when you introduce a custom `notifications-secret` header that bypasses the whole backstage authentication process.